This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [1]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = ''

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [2]:
# First, import the relevant modules
import requests
import json
from collections import defaultdict

In [3]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
url = 'https://www.quandl.com/api/v3/datasets/FSE/AFX_X/data.json'
req = requests.get(url + '?api_key=' + API_KEY + '&start_date=2020-01-02&end_date=2020-01-02')
json_data = req.json()

In [4]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
json_data

{'dataset_data': {'limit': None,
  'transform': None,
  'column_index': None,
  'column_names': ['Date',
   'Open',
   'High',
   'Low',
   'Close',
   'Change',
   'Traded Volume',
   'Turnover',
   'Last Price of the Day',
   'Daily Traded Units',
   'Daily Turnover'],
  'start_date': '2020-01-02',
  'end_date': '2020-01-02',
  'frequency': 'daily',
  'data': [['2020-01-02',
    114.0,
    114.9,
    112.7,
    113.8,
    None,
    143464.0,
    16291666.0,
    None,
    None,
    None]],
  'collapse': None,
  'order': None}}

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [5]:
##FSE, AFX_X, entire 2017
req_2017 = requests.get(url + '?api_key=' + API_KEY + '&start_date=2017-01-01&end_date=2017-12-31')

##Convert JSON object to Python dictionary
data_2017 = req_2017.json()
#data_2017

In [6]:
##Question 3
daily = data_2017.get('dataset_data').get('data')
daily[:5]

[['2017-12-29',
  51.76,
  51.94,
  51.45,
  51.76,
  None,
  34640.0,
  1792304.0,
  None,
  None,
  None],
 ['2017-12-28',
  51.65,
  51.82,
  51.43,
  51.6,
  None,
  40660.0,
  2099024.0,
  None,
  None,
  None],
 ['2017-12-27',
  51.45,
  51.89,
  50.76,
  51.82,
  None,
  57452.0,
  2957018.0,
  None,
  None,
  None],
 ['2017-12-22',
  51.05,
  51.5,
  50.92,
  51.32,
  None,
  71165.0,
  3641949.0,
  None,
  None,
  None],
 ['2017-12-21',
  51.16,
  51.52,
  50.9,
  51.4,
  None,
  120649.0,
  6179433.0,
  None,
  None,
  None]]

In [7]:
open_prices = []
for x in daily:
    if x[1] != None:
        open_prices.append((x[1], x[0]))
max(open_prices)

(53.11, '2017-12-14')

In [8]:
min(open_prices)

(34.0, '2017-01-24')

The highest opening price was \\$53.11 on 2017-12-14 and the lowest opening price was \\$34.00 on 2017-01-24.

In [9]:
##Question 4
delta_prices = []
for x in daily:
    if x[2] != None and x[3] != None:
        delta_prices.append((x[2] - x[3], x[0]))
max(delta_prices)

(2.8100000000000023, '2017-05-11')

The greatest difference in prices in a day was \\$2.81 on 2017-05-11

In [10]:
##Question 5
close_prices = []
for x in daily:
    if x[4] != None:
        close_prices.append((x[4], x[0]))
min(close_prices)

(34.06, '2017-01-23')

In [11]:
max(close_prices)

(53.09, '2017-12-13')

In [12]:
max(close_prices)[0]-min(close_prices)[0]

19.03

Highest closing price was \\$53.09 on 2017-12-13 and lowest closing price was \\$34.06 on 2017-01-23 for a difference of \\$19.03

In [13]:
##Question 6
trading_volume = []
for x in daily:
    if x[6] != None:
        trading_volume.append(x[6])
annual_avg = sum([x for x in trading_volume])/len(trading_volume)
annual_avg

89124.33725490196

The annual average of trading volume in 2017 was 89124.

In [14]:
##Question 7
len(trading_volume)

255

In [15]:
trading_volume.sort()
trading_volume[:5]

[45.0, 112.0, 290.0, 522.0, 587.0]

In [16]:
trading_volume[len(trading_volume)//2]

76286.0

In [17]:
trading_volume_dated = []
for x in daily:
    if x[6] != None:
        trading_volume_dated.append((x[6], x[0]))

In [18]:
for x, y in trading_volume_dated:
    if x == 76286.0:
        print(y)

2017-11-13


The median trading volume was 76286 and occurred on 2017-11-13.